core/src/main/java/org/apache/iceberg/DeleteFileIndex.java

```diff
@@ -0,0 +1,270 @@
+      ListMultimap<Pair<Integer, StructLikeWrapper>, ManifestEntry<DeleteFile>> deleteFilesByPartition =
+          Multimaps.newListMultimap(Maps.newHashMap(), Lists::newArrayList);
+      for (Pair<Integer, ManifestEntry<DeleteFile>> specIdAndEntry : deleteEntries) {
+        int specId = specIdAndEntry.first();
+        ManifestEntry<DeleteFile> entry = specIdAndEntry.second();
+        deleteFilesByPartition.put(Pair.of(specId, StructLikeWrapper.wrap(entry.file().partition())), entry);
+      }
+
+      // sort the entries in each map value by sequence number and split into sequence numbers and delete files lists
+      Map<Pair<Integer, StructLikeWrapper>, Pair<long[], DeleteFile[]>> sortedDeletesByPartition = Maps.newHashMap();
+      // also, separate out equality deletes in an unpartitioned spec that should be applied globally
+      long[] globalApplySeqs = null;
+      DeleteFile[] globalDeletes = null;
+      for (Pair<Integer, StructLikeWrapper> partition : deleteFilesByPartition.keySet()) {
+        if (specsById.get(partition.first()).isUnpartitioned()) {
+          Preconditions.checkState(globalDeletes == null, "Detected multiple partition specs with no partitions");
+
+          List<Pair<Long, DeleteFile>> eqFilesSortedBySeq = deleteFilesByPartition.get(partition).stream()
+              .filter(entry -> entry.file().content() == FileContent.EQUALITY_DELETES)
+              .map(entry ->
+                  // a delete file is indexed by the sequence number it should be applied to
+                  Pair.of(entry.sequenceNumber() - 1, entry.file()))
+              .sorted(Comparator.comparingLong(Pair::first))
+              .collect(Collectors.toList());
+
+          globalApplySeqs = eqFilesSortedBySeq.stream().mapToLong(Pair::first).toArray();
+          globalDeletes = eqFilesSortedBySeq.stream().map(Pair::second).toArray(DeleteFile[]::new);
+
+          List<Pair<Long, DeleteFile>> posFilesSortedBySeq = deleteFilesByPartition.get(partition).stream()
+              .filter(entry -> entry.file().content() == FileContent.POSITION_DELETES)
+              .map(entry -> Pair.of(entry.sequenceNumber(), entry.file()))
+              .sorted(Comparator.comparingLong(Pair::first))
+              .collect(Collectors.toList());
+
+          long[] seqs = posFilesSortedBySeq.stream().mapToLong(Pair::first).toArray();
+          DeleteFile[] files = posFilesSortedBySeq.stream().map(Pair::second).toArray(DeleteFile[]::new);
+
+          sortedDeletesByPartition.put(partition, Pair.of(seqs, files));
+
+        } else {
+          List<Pair<Long, DeleteFile>> filesSortedBySeq = deleteFilesByPartition.get(partition).stream()
+              .map(entry -> {
+                // a delete file is indexed by the sequence number it should be applied to
+                long applySeq = entry.sequenceNumber() -
+                    (entry.file().content() == FileContent.EQUALITY_DELETES ? 1 : 0);
```

Comment:
Why do we treat positional and equality deletes differently? Is it because the equality delete should not delete data in the snapshot it was added?

Reply:
Exactly.

If you want to replace a row with an equality delete, then the delete must apply to data before the snapshot is committed. Otherwise, you'd delete the row that you're writing as a replacement because the delete is not targeted to a file/position within a partition.

We can support same-sequence-number deletes for positional deletes because they target specific rows in files. And, supporting this is how we can delete within a checkpoint for streaming use cases when we encounter the same record twice.